exec/java-exec/src/main/java/org/apache/drill/exec/expr/fn/impl/SqlPatternContainsMatcher.java

```diff
@@ -19,44 +19,283 @@
 
 import io.netty.buffer.DrillBuf;
 
-public class SqlPatternContainsMatcher extends AbstractSqlPatternMatcher {
+/** SQL Pattern Contains implementation */
+public final class SqlPatternContainsMatcher extends AbstractSqlPatternMatcher {
+  private final MatcherFcn matcherFcn;
 
   public SqlPatternContainsMatcher(String patternString) {
     super(patternString);
+
+    // Pattern matching is 1) a CPU intensive operation and 2) pattern and input dependent. The conclusion is
+    // that there is no single implementation that can do it all well. So, we use multiple implementations
+    // chosen based on the pattern length.
+    if (patternLength == 1) {
+      matcherFcn = new Matcher1();
+    } else if (patternLength == 2) {
+      matcherFcn = new Matcher2();
+    } else if (patternLength == 3) {
+      matcherFcn = new Matcher3();
+    } else if (patternLength < 10) {
+      matcherFcn = new MatcherN();
+    } else {
+      matcherFcn = new BoyerMooreMatcher();
+    }
   }
 
   @Override
   public int match(int start, int end, DrillBuf drillBuf) {
+    return matcherFcn.match(start, end, drillBuf);
+  }
+
+  //--------------------------------------------------------------------------
+  // Inner Data Structure
+  // --------------------------------------------------------------------------
+
+  /** Abstract matcher class to allow us pick the most efficient implementation */
+  private abstract class MatcherFcn {
+    protected final byte[] patternArray;
+
+    protected MatcherFcn() {
+      assert patternByteBuffer.hasArray();
+
+      patternArray = patternByteBuffer.array();
+    }
+
+    /**
+     * @return 1 if the pattern was matched; 0 otherwise
+     */
+    protected abstract int match(int start, int end, DrillBuf drillBuf);
+  }
+
+  /** Handles patterns with length one */
+  private final class Matcher1 extends MatcherFcn {
 
-    if (patternLength == 0) { // Everything should match for null pattern string
-      return 1;
+    private Matcher1() {
+      super();
     }
 
-    final int txtLength = end - start;
+    /** {@inheritDoc} */
+    @Override
+    protected final int match(int start, int end, DrillBuf drillBuf) {
+      final int lengthToProcess = end - start;
+      final byte firstPattByte  = patternArray[0];
 
-    // no match if input string length is less than pattern length
-    if (txtLength < patternLength) {
+      // simplePattern string has meta characters i.e % and _ and escape characters removed.
+      // so, we can just directly compare.
+      for (int idx = 0; idx < lengthToProcess; idx++) {
+        byte inputByte = drillBuf.getByte(start + idx);
+
+        if (firstPattByte != inputByte) {
```

Comment:
These optimizations are useless with modern compilers. This would have been useful if the inputByte value was the same.. then we would save few instructions.